Update dead compaction detector to handle metadata/root

cshannon opened the pull request:
This updates DeadCompactionDetector to also look at the root and metadata table metadata for dead compactions on those tables

This closes #4340

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/DeadCompactionDetector.java:
```diff
@@ -145,9 +146,14 @@ private void detectDeadCompactions() {
+        // Listing the data levels vs using DataLevel.values() prevents unexpected
+        // behavior if a new DataLevel is added
+        .of(DataLevel.ROOT, DataLevel.METADATA, DataLevel.USER)
+        .map(dataLevel -> context.getAmple().readTablets().forLevel(dataLevel)
```
Uncertain if this will close the Closeable returned by context.getAmple().readTablets()

cshannon replied:
This is addressed in #4357

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/DeadCompactionDetector.java:
```diff
@@ -145,9 +146,14 @@ private void detectDeadCompactions() {
-    try (TabletsMetadata tabletsMetadata = context.getAmple().readTablets().forLevel(DataLevel.USER)
-        .filter(new HasExternalCompactionsFilter()).fetch(ColumnType.ECOMP, ColumnType.PREV_ROW)
-        .build()) {
+    try (Stream<TabletMetadata> tabletsMetadata = Stream
```
A few lines up there is a TODO that can be removed

keith-turner reviewed (commented):
The conditional tablet mutator [does not support](https://github.com/apache/accumulo/blob/3fe11150145fab6d920aae8582340ad7e07a8fba/server/base/src/main/java/org/apache/accumulo/server/metadata/ConditionalTabletsMutatorImpl.java#L75-L80) mutation multiple data levels.  It used in [this code](https://github.com/apache/accumulo/blob/3fe11150145fab6d920aae8582340ad7e07a8fba/server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/CompactionCoordinator.java#L719) called by the dead compaction detector.  

One option for dealing with this is to have the code partition the extents by data level and use a conditional writer per partition.  Another option is to change impl of conditional tablets mutator to support multiple levels.

May be good to test collection of dead compactions at multiple levels at the same time.

cshannon commented:
@keith-turner - I pushed an update which groups the compactions by data level when processing with the mutator since as you pointed out, the conditional mutator does not support multiple data levels. I also refactored the tests and I added a new test that creates dead compactions for all 3 levels at the same time to verify it works. I verified the new test failed before my change and now works with the grouping.

cshannon commented:
I should also add that I chose to just partition the extents into buckets at the end when processing the remaining compactions that were considered as dead and passed to the mutator. It made things a lot simpler as the rest of the code didn't have to deal with 3 different buckets/groups and there should be minimal performance hit because my assumption is dead compactions should not be very common so there shouldn't be many items to have to process and partition.

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/compaction/coordinator/CompactionCoordinator.java:
```diff
@@ -710,6 +715,15 @@ public void compactionFailed(TInfo tinfo, TCredentials credentials, String exter
     compactionFailed(Map.of(ecid, KeyExtent.fromThrift(extent)));
   }
 
+  void compactionsFailed(Map<ExternalCompactionId,Pair<KeyExtent,DataLevel>> compactionsByLevel) {
+    // Need to process each level by itself because the conditional tablet mutator does not support
+    // mutating multiple data levels at the same time
+    compactionsByLevel.entrySet().stream()
+        .collect(groupingBy(entry -> entry.getValue().getSecond(),
+            Collectors.toMap(Entry::getKey, e -> e.getValue().getFirst())))
+        .forEach((level, compactions) -> compactionFailed(compactions));
+  }
+
```
Could avoid tracking the level in the outer code by computing it here when partitioning.  Would need to rename the function that takes a map of data for a level.  Renamed it to `compactionFailedForLevel` in the suggestion.

```suggestion
  void compactionsFailed(Map<ExternalCompactionId,KeyExtent> compactionsByLevel) {
    // Need to process each level by itself because the conditional tablet mutator does not support
    // mutating multiple data levels at the same time
    compactionsByLevel.entrySet().stream()
        .collect(groupingBy(entry -> DataLevel.of(entry.getValue().tableId()),
            Collectors.toMap(Entry::getKey, Entry::getValue)))
        .forEach((level, compactions) -> compactionFailedForLevel(compactions));
  }
```

cshannon replied:
Thanks for the suggestion, that should simplify things. I can update that and merge later this week when I get some time, probably Wednesday.
